_source/_posts/2019-05-30-okta-token-hooks.md

```diff
@@ -0,0 +1,395 @@
+
+You'll first set up an OIDC application in Okta and be able to see the types of tokens you get back. Then, you'll configure a Spring Boot API app that allows you to register you're favorite beers (this is based on [previous post]() from my friend and colleague, Matt Raible, who is really into beers). This Spring Boot app also has API endpoints to register token hooks with Okta and to service token patch requests from Okta. Finally, you'll see that your favorite beers are now included in the payload of a token thanks to the hooks you registered.
+
+To get ready for the hands-on parts of this post, set yourself up with the following:
+
+* Create a free Okta developer org at [https://developer.okta.com/signup/](https://developer.okta.com/signup/)
+* Install [HTTPie](https://httpie.org), a modern curl replacement for interacting with API endpoints later
+* Install [ngrok](https://ngrok.com/), an app that allows public access to a locally running server
+* Optional: Create a free Heroku account at [https://signup.heroku.com/](https://signup.heroku.com/)
+
+To start, let's take a step back and introduce OIDC and OAuth.
+
+## Three Minute Overview of OpenID Connect and OAuth 2.0
+
+<!-- // TODO - This is copied from a previous post verbatim. Should I change? // -->
+
+In the beginning there were siloed web sites that didn't talk to each other, and it was sad.
+
+Sites like Yelp started wanting access to the contact information you had in your Google Contacts. So, Yelp naturally collected your Google username and password so that it could access your contacts. You gave Yelp your permission, so this was all good, Yes? No! With your username and password, Yelp could access your email, your docs - everything you had in Google - not just your contacts. And, worse, Yelp had to store your password in a way that it could use it in plaintext and there was no standard way to revoke your consent to Yelp to access your Google account.
+
+We needed an authorization framework that would allow you to grant access to certain information without you giving up your password. Cue OAuth.
+
+Three revisions later, we're at OAuth 2.0 (there was 1.0 and 1.0a before it) and all's right with the world. Now, an application like Yelp (a `Client Application`) can request an `Access Token` from a service like Google (an `Authorization Server`). You (the `Resource Owner`) log into Google with your credentials and give your `Consent` to Yelp to access your contacts (and only your contacts). `Access Token` in hand, Yelp makes a request of the Google Contacts API (the `Resource Server`) and gets your contacts. Yelp never sees your password and never has access to anything more than you've consented to. And, you can withdraw your consent at any time.
+
```

Comment:
Yelp never sees your password and never has access to anything more than you've consented to. And, you can withdraw your consent at any time. => Yelp never sees your password and never has access to anything more than you've consented to, and you can withdraw your consent at any time.